apps/prairielearn/src/lib/assessment.js

```diff
@@ -36,486 +38,429 @@ const InstanceLogSchema = z.object({
+      gradable,
+      false, // `broken`
+      null, // `params`
+      null, // `true_answer`
+      content.grading.feedback,
+      {}, // `partial_scores`
+      content.grading.score,
+      null, // `v2_score`: gross legacy, this can safely be null
+    ]);
+    const assessment_instance_id = await sqldb.queryRow(
+      sql.select_assessment_for_grading_job,
+      { grading_job_id: content.gradingId },
+      IdSchema,
     );
-  },
+    await promisify(ltiOutcomes.updateScore)(assessment_instance_id);
+  } finally {
+    externalGradingSocket.gradingJobStatusUpdated(content.gradingId);
+  }
+}
+
+/**
+ * Updates statistics for all assessments in a course instance, but only if an
+ * update is needed.
+ *
+ * @param {string} course_instance_id - The course instance ID.
+ * @returns {Promise<void>}
+ */
+export async function updateAssessmentStatisticsForCourseInstance(course_instance_id) {
+  const rows = await sqldb.queryRows(
+    sql.select_assessments_for_statistics_update,
+    { course_instance_id },
+    IdSchema,
+  );
+  await async.eachLimit(rows, 3, updateAssessmentStatistics);
+}
 
-  async selectAssessmentInstanceLogCursor(assessment_instance_id, include_files) {
-    return sqldb.queryValidatedCursor(
-      sql.assessment_instance_log,
-      { assessment_instance_id, include_files },
-      InstanceLogSchema,
+/**
+ * Updates statistics for an assessment, if needed.
+ *
+ * @param {string} assessment_id - The assessment ID.
+ * @returns {Promise<void>}
+ */
+export async function updateAssessmentStatistics(assessment_id) {
+  await sqldb.runInTransactionAsync(async () => {
+    // lock the assessment
+    await sqldb.queryOneRowAsync(sql.select_assessment_lock, { assessment_id });
+
+    // check whether we need to update the statistics
+    const needs_statistics_update = await sqldb.queryRow(
+      sql.select_assessment_needs_statisics_update,
+      { assessment_id },
+      z.boolean(),
     );
-  },
-};
+    if (!needs_statistics_update) return;
+
+    // update the statistics
+    await sqldb.queryOneRowAsync(sql.update_assessment_statisics, { assessment_id });
+  });
+}
+
+/**
+ * Selects a log of all events associated to an assessment instance.
+ *
+ * @param {string} assessment_instance_id - The ID of the assessment instance.
+ * @param {boolean} include_files - A flag indicating if submitted files should be included in the
+ * log.
+ * @returns {Promise<Array<z.infer<typeof InstanceLogSchema>>>} the results of the log query.
+ */
+export async function selectAssessmentInstanceLog(assessment_instance_id, include_files) {
+  return sqldb.queryRows(
+    sql.assessment_instance_log,
+    { assessment_instance_id, include_files },
+    InstanceLogSchema,
+  );
+}
+
+/**
+ *
+ * @param {string} assessment_instance_id
+ * @param {boolean} include_files
+ * returns {Promise<sqldb.CursorIterator<z.infer<typeof InstanceLogSchema>>>} (Not using jsdoc since CursorIterator is not exported yet)
```

Comment:
One of us should follow up to export this type.

Reply:
Will do when I convert it to TS.